ompi/info/info_memkind.c

```diff
@@ -0,0 +1,545 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (c) 2024      Advanced Micro Devices, Inc. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "info/info_memkind.h"
+#include "opal/util/argv.h"
+#include "opal/mca/accelerator/accelerator.h"
+#include "opal/mca/accelerator/base/base.h"
+#include "opal/util/printf.h"
+#include "ompi/errhandler/errcode.h"
+#include "ompi/constants.h"
+
+static opal_mutex_t ompi_info_memkind_mutex = OPAL_MUTEX_STATIC_INIT;
+static ompi_memkind_t *ompi_info_memkind_available;
+static int ompi_info_memkind_num_available = 0;
+
+#define FREE_STR_ARR(_str_arr) { \
+    int _i=0;                    \
+    char *_t = _str_arr[_i];     \
+    while (NULL != _t) {         \
+        free (_t);               \
+        _t = _str_arr[++_i];     \
+    }                            \
+    free (_str_arr);             \
+}
+
+static int ompi_info_memkind_get_num_unique (char **memkind_strs, int num_max)
+{
+    int iter = 0, pos = 0;
+    int num_unique = 0;
+    char **tmp_arr;
+
+    tmp_arr = (char**) malloc ( (num_max + 1) * sizeof(char*));
+
+    char *m = memkind_strs[iter];
+    while (m != NULL) {
+        char **tmp_str = opal_argv_split (m, ':');
+        tmp_arr[pos++] = strdup (tmp_str[0]);
+
+        FREE_STR_ARR(tmp_str);
+        m = memkind_strs[++iter];
+    }
+    tmp_arr[pos] = NULL;
+
+    iter = 0;
+    m = tmp_arr[iter];
+    while (NULL != m) {
+        bool already_seen = false;
+
+        for (int i = 0; i < iter; i++) {
+            if (!strncmp(m, tmp_arr[i], strlen(m))) {
+                already_seen = true;
+                break;
+            }
+        }
+        if (!already_seen)  {
+            num_unique++;
+        }
+        m = tmp_arr[++iter];
+    }
+    FREE_STR_ARR(tmp_arr);
+
+    return num_unique;
+}
+
+#if 0
```

Comment:
do we want to leave this debug code in the PR?

Reply:
I would like to keep the code in there, I am 99.9% sure that we will have use for that in the near future. Furthermore, there are other components/code parts that also have dump functionality in the code base. The `#if 0` is just to get rid of compiler warning about unused static function